packages/components/src/Lettera/Models.purs

```diff
@@ -633,3 +639,105 @@ editorialIdToUuid editorialId =
 url_namespace :: UUID.UUID
 url_namespace = unsafePartial $ fromJust $ UUID.parseUUID "6ba7b811-9dad-11d1-80b4-00c04fd430c8"
 
+encodeRequestHeaders :: Array AX.RequestHeader -> Maybe Json
+encodeRequestHeaders [] = Nothing
+encodeRequestHeaders hs =
+  let encodeHeader h obj = flip extend obj $ case h of
+        AX.Accept (MediaType v)      -> "Accept" := v
+        AX.ContentType (MediaType v) -> "ContentType" := v
+        AX.RequestHeader k v         -> k := v
+  in Just $ foldr encodeHeader jsonEmptyObject hs
+
+encodeRequest :: forall a . AX.Request a -> Json
+encodeRequest x =
+  let method = case x.method of
+        Left m  -> encodeJson (show m)
+        Right s -> encodeJson (show s)
+  in "method"  := method ~>
+     "url"     := x.url  ~>
+     "headers" :=? encodeRequestHeaders x.headers ~>?
+     -- Currently we're only using GET requests so we don't need content:
+     -- "content" := x.content ~>
+     jsonEmptyObject
+
+encodeResponseHeaders :: Array AX.ResponseHeader -> Json
+encodeResponseHeaders hs =
+  let encodeHeader (AX.ResponseHeader k v) obj = k := v ~> obj
+  in foldr encodeHeader jsonEmptyObject hs
+
+encodeResponse :: forall a . AX.Response a -> Json
+encodeResponse x =
+  let statusCode = case x.status of
+        AX.StatusCode n -> n
+  in "statusCode" := statusCode ~>
+--     "statusText" := x.statusText ~>
+     "headers" := encodeResponseHeaders x.headers ~>
+     jsonEmptyObject
+
+encodeError :: String -> Maybe Json -> Maybe Json -> Maybe Json -> Json
+encodeError name req err code =
+  "err" := name ~>
+  (("req" := _) <$> req) ~>?
+  (("msg" := _) <$> err) ~>?
+  (("code" := _) <$> code) ~>?
+  jsonEmptyObject
+
+instance encodeJsonLetteraError :: EncodeJson LetteraError where
+  encodeJson e = case e of
+    ResponseError { letteraErrorRequest, letteraErrorError } ->
+      encodeError "ResponseError"
+                  (Just letteraErrorRequest)
+                  (Just (encodeString (AX.printError letteraErrorError)))
+                  Nothing
+    HttpError { letteraErrorRequest, letteraErrorResponse, letteraErrorCode } ->
+      encodeError "HttpError"
+                  (Just letteraErrorRequest)
+                  (Just letteraErrorResponse)
+                  (Just (encodeInt letteraErrorCode))
+    ParseError ->
+      encodeError "ParseErrror" Nothing Nothing Nothing
+
+data LetteraError
+  = ResponseError
+    { letteraErrorRequest  :: Json
```

Comment:
I think I said it already but don't use `Json` on this level. Leave generating it to the `encodeJson` instance. Ditto for `HttpError`. If we need users of this code to do actions based on responses then that's a lot more convenient format.

Reply:
Okay, but then we'll have more polymorphic type variables to deal with.

Reply:
I started to think that maybe including requests and responses in the error type is not the best solution to start with. That information is needed for logging but is it really an essential part of an error value? Perhaps we could consider moving the logging from Mosaico to Affresco to the place where we have the best information about the requests, responses and/or other context?

Reply:
It seems that If I generalise the `Json` type occurrences into type variables, then Mosaico and Affresco disagree whether the `getFrontpage` and `getFrontpageHtml` are dealing with `LetteraResponse String String a` or `LetteraResponse Json Json a`.